### PR TITLE
[Perf] Async logits proc and sampling metadata state update

### PR DESCRIPTION
**Pull request by njhill**
Mentioned by @WoosukKwon in https://github.com/vllm-project/vllm/issues/23233.

The work to update logits processor states and construct the sampling metadata be deferred in model runner `execute_model` to overlap CPU with forward pass.

This has an advantage of taking the logits processor `update_states` calls off the critical path, and means we can trade off a bit of efficiency for simplicity in the API design.

There should be some incremental benefit as we migrate hard-wired sampling parameters into logits processors since their state updates are still happening in `_update_states()` before the forward pass.

### CODE REVIEW

File: vllm/v1/worker/gpu_model_runner.py [hunk 1]
```diff
@@ -495,7 +488,8 @@ def _update_states(self, scheduler_output: "SchedulerOutput") -> None:
                 mm_positions=new_req_data.mm_positions,
                 sampling_params=sampling_params,
                 pooling_params=pooling_params,
-                generator=generator,
+                # generator is populated in _update_sampling_states()
```
Comment (robertgshaw2-redhat) [inline vllm/v1/worker/gpu_model_runner.py:491]: I don't think its worth delaying this construction. This only happens once per request so I don't think the code readability change is worth the per win

**Comment by robertgshaw2-redhat** [comment]
nice idea!

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:491]: I guess it didn't really seem less readable to me - the generator is for seeded sampling so in some way fits naturally in `_update_sampling_states()`...

File: vllm/v1/worker/gpu_model_runner.py [hunk 2]
```diff
@@ -1651,6 +1666,10 @@ def execute_model(
             assert model_output_broadcast_data is not None
             logits = model_output_broadcast_data["logits"]
 
+        # Apply batch state updates required for sampling stage;
+        # can run in parallel with model forward-pass.
+        self._update_sampling_states(scheduler_output)
```
Comment (WoosukKwon) [inline vllm/v1/worker/gpu_model_runner.py:1671]: @njhill Doesn't this cause CPU-GPU synchronization?

File: vllm/v1/worker/gpu_model_runner.py (hunk 1 above)
Comment (WoosukKwon) [reply vllm/v1/worker/gpu_model_runner.py:491]: I agree with @robertgshaw2-redhat
I think it's more natural to initialize it here.

File: vllm/v1/worker/gpu_model_runner.py (hunk 2 above)
Comment (njhill) [reply vllm/v1/worker/gpu_model_runner.py:1671]: @WoosukKwon I don't think so, only non-blocking CPU->GPU of tensors that aren't used until the sampler...

Sync would only be for GPU->CPU right?
